nexus_motion_planner/src/test_motion_plan_cache.cpp

```diff
@@ -0,0 +1,1165 @@
+
+  check_and_emit(
+    cache->count_plans(g_robot_name) == 1, prefix, "One plan in cache");
+
+  // Fetch matching, no tolerance
+  //
+  // Exact key match should have cache hit
+  prefix = "test_motion_plans.put_first.fetch_matching_no_tolerance";
+
+  auto fetched_plans = cache->fetch_all_matching_plans(
+    *move_group, g_robot_name, plan_req, 0, 0);
+
+  auto fetched_plan = cache->fetch_best_matching_plan(
+    *move_group, g_robot_name, plan_req, 0, 0);
+
+  check_and_emit(fetched_plans.size() == 1, prefix, "Fetch all returns one");
+  check_and_emit(
+    fetched_plan != nullptr, prefix, "Fetch best plan is not nullptr");
+
+  check_and_emit(
+    *fetched_plans.at(0) == *fetched_plan,
+    prefix, "Fetched plan on both fetches match");
+
+  check_and_emit(
+    *fetched_plan == plan, prefix, "Fetched plan matches original");
+
+  check_and_emit(
+    fetched_plan->lookupDouble("execution_time_s") == execution_time,
+    prefix, "Fetched plan has correct execution time");
+
+  check_and_emit(
+    fetched_plan->lookupDouble("planning_time_s") == planning_time,
+    prefix, "Fetched plan has correct planning time");
+
+  // Fetch with is_diff
+  //
+  // is_diff key should be equivalent to exact match if robot state did not
+  // change, hence should have cache hit
+  prefix = "test_motion_plans.put_first.fetch_is_diff_no_tolerance";
+
+  fetched_plans = cache->fetch_all_matching_plans(
+    *move_group, g_robot_name, is_diff_plan_req, 0, 0);
+
+  fetched_plan = cache->fetch_best_matching_plan(
+    *move_group, g_robot_name, is_diff_plan_req, 0, 0);
+
+  check_and_emit(fetched_plans.size() == 1, prefix, "Fetch all returns one");
+  check_and_emit(
+    fetched_plan != nullptr, prefix, "Fetch best plan is not nullptr");
+
+  check_and_emit(
+    *fetched_plans.at(0) == *fetched_plan,
+    prefix, "Fetched plan on both fetches match");
+
+  check_and_emit(
+    *fetched_plan == plan, prefix, "Fetched plan matches original");
+
+  check_and_emit(
+    fetched_plan->lookupDouble("execution_time_s") == execution_time,
+    prefix, "Fetched plan has correct execution time");
+
+  check_and_emit(
+    fetched_plan->lookupDouble("planning_time_s") == planning_time,
+    prefix, "Fetched plan has correct planning time");
+
+  // Fetch non-matching, out of tolerance
+  //
+  // Non-matching key should not have cache hit
+  prefix = "test_motion_plans.put_first.fetch_non_matching_out_of_tolerance";
+
+  fetched_plans = cache->fetch_all_matching_plans(
+    *move_group, g_robot_name, close_matching_plan_req, 0, 0);
+
+  fetched_plan = cache->fetch_best_matching_plan(
+    *move_group, g_robot_name, close_matching_plan_req, 0, 0);
+
+  check_and_emit(fetched_plans.size() == 0, prefix, "Fetch all returns empty");
+  check_and_emit(
+    fetched_plan == nullptr, prefix, "Fetch best plan is nullptr");
+
+  // Fetch non-matching, in tolerance
+  //
+  // Close key within tolerance limit should have cache hit
```

Comment:
Add test where only one of start or goal tolerance is within tolerance.

Reply:
a235d207d1a90b0e13a6ef7d2032568437971266